src/library/scala/collection/View.scala

```diff
@@ -404,9 +404,15 @@ object View extends IterableFactory[View] {
 
   @SerialVersionUID(3L)
   private[collection] class Patched[A](underlying: SomeIterableOps[A], from: Int, other: IterableOnce[A], replaced: Int) extends AbstractView[A] {
-    def iterator: Iterator[A] = underlying.iterator.patch(from, other.iterator, replaced)
+    // we may be unable to traverse `other` more than once, so we need to cache it if that's the case
```

Comment:
In which situation may we traverse `other` more than once?

Reply:
in principle, any view can be traversed arbitrarily many times, so `coll.view.patched(1, Iterator.single(1), 1)` could produce an incorrect result without this change if traversed twice (which is permitted)

Reply:
I think what caused this issue to crop up for me was the change to call `size` on `Iterable`s in `ArrayBuffer#addAll`

Reply:
incidentally, you could already hit this bug using `insertAll`:

```scala
scala> ArrayBuffer(1)
     |   .insertAll(
     |     1,
     |     List(2)
     |       .view
     |       .patch(1, List(3, 4, 5).iterator, 0)
     |   )
java.util.NoSuchElementException: head of empty list
  at scala.collection.immutable.Nil$.head(List.scala:629)
  at scala.collection.immutable.Nil$.head(List.scala:628)
  at scala.collection.StrictOptimizedLinearSeqOps$$anon$1.next(LinearSeq.scala:253)
  at scala.collection.Iterator$$anon$17.next(Iterator.scala:915)
  at scala.collection.mutable.ArrayBuffer.insertAll(ArrayBuffer.scala:181)
  ... 32 elided

scala> ArrayBuffer(1)
     |   .insertAll(
     |     1,
     |     List(2)
     |       .view
     |       .patch(1, List(3, 4, 5), 0)
     |   )

scala>
```

Reply:
sure

Reply:
there's a simpler repro that we can use as a test, even.

```scala
scala> val view = List(2).view.patch(1, List(3, 4, 5).iterator, 0)
val view: scala.collection.View[Int] = View(<not computed>)

scala> view.toList
val res0: List[Int] = List(2, 3, 4, 5)

scala> view.toList
val res1: List[Int] = List(2)
```